### ISSUE TITLE
icecc-scheduler - Option to allow distributed file linking

### ISSUE THREAD

**Issue by MrD2** [icecc/icecream#482]
In a distributed environment, if we use an attached network storage maybe its possible to link files also in the workers? 
This would release the scheduler to only do, well scheduling.

I have a working example (crude) of icecc working on Kubernetes cluster, scaling worker pods by load. If the linking is possible to be done also by the worker nodes it would make a truly scalable solution.

Also, as the icecc has LB (contrary to what distcc does) we can even run such compilation jobs as a low priority preemptible containers scenario.

**Comment by llunak** [comment]
What is that supposed to improve? Especially with debug info the files to link and the resulting binary are possibly going to be huge and as such not worth doing linking remotely. And generally linking is not done that often and that in parallel.

If you find linking slow, you probably actually want a better linker (GNU gold or LLVM lld can both replace the sluggish BFD linker that is often used by default).

**Comment by MrD2** [comment]
I don't know much about compiler. 

But if imagining i'm building the code from a distributed file system. Is it possible to do this distributed linking? The idea is that i can have the scheduler only to schedule stuff, and the workers to actually do the jobs (all of them).
Even if it takes more time (per job), or repeated builds, its probably faster if i have the resources to scale the job no 100/200/300 parallel processes

**Comment by MrD2** [comment]
Do you have a doxygen document with graphvz for this project available online?

**Comment by llunak** [comment]
Icecream does not use distributed file systems, but even if support for it is added, I still don't see a difference between Icecream transferring the data or the filesystem doing it - it still seems not worth it compared to the cost of linking locally. You're welcome to try it yourself if you can show an improvement that way, but without that I'm refusing the idea as not worth it, sorry.

**Comment by MrD2** [comment]
The data on the icecream is stored all in one place.

So all the workers and the scheduler would have the same bin dir tree (admitting no concurrency)
This would allow the workers to also link the files, as all the processed data would be available in one unique place

**Comment by llunak** [comment]
I'm still confused about the setup, but either way this request still doesn't make sense to me. If that one place to store the data is a network drive, then you'll probably get the speed improvement by using a local disk, and then the distributed linking idea is not needed. Or, if by any chance that one place is a local storage and this somehow involves virtualization on that one machine, then just don't complicate your life and build everything directly on the machine.
